Clients: Rich command-line client outputs #6987

ericbanzuzi opened the pull request:
For more details, view issue #6987.

###  Key changes:
-  Adds rich as a dependency
- Adds the CLI outputs using Rich for all the commands in `bin/rucio` and `bin/rucio-admin`.
- Adds a configuration option to choose between Rich or tabulate for CLI output.  Users can specify this in the config file by adding `cli = rich` in the `[experimental]` section. If the option `cli` does not exist or has a value other than `rich`, the tabulate CLI will be used by default.
- Implemented paging for outputs (`less -FRSXMK`) when the output exceeds the terminal size.
- Option `--no-pager ` to disable paging.

### Pager:

The paging is now done with `pydoc`, similarly as rich `console.pager()` does it. Here the steps of `console.pager()` is done manually so that the the command methods return the `SUCCESS` code before displaying the output and waiting for pager response. Rich only allows to use `console.pager()` as through a context manager, meaning that the method would only return `SUCCESS` when the user exits the pager.

### Dimming colors in rows:

I've decided to only apply dimming to have alternating coloring of rows only to the tables in which the numbe rof entries in unknown beforehand. This would mean that it is applied to all tables except the ones showing key-value pairs. The commands without alternating colors are: `rucio whoami`, `rucio get-metadata`, `rucio stat`, ` rucio list-rse-attributes`, `rucio rule-info`, `rucio-admin subscription list`, `rucio-admin account info`,  `rucio-admin rse get-attribute` and `attributes` `settings` `protocols` in `rucio-admin rse info`. I felt like dimming in the key-value tables was not such a natural effect.


### Styles:

I tried to choose a colour palette that highlights states with semantic colours to indicate whether something is positive, negative or neutral. For other types, the chosen colours are contrast colours, aiming to give clear separability without imposing emotional connotations. The chosen colours in a default mac terminal can be seen below:

<img width="215" alt="cli_palette" src="https://github.com/user-attachments/assets/0f52fc95-0622-4be7-a844-b1daaf434d59">
<img width="197" alt="cli_plaette_on_white" src="https://github.com/user-attachments/assets/7c1dc3f0-9f6e-4b0f-8379-002e05bfc591">

They might differ slightly in other terminal applications.

dchristidis reviewed (changes requested):
Very promising! I have two major comments that should be addressed first:

 1. It’s not beneficial to have identical definitions in both `rucio` and `rucio-admin`. Please move to `lib/rucio/richclient.py` and import as necessary.
 2. All new functions should have type hints.

There will be more comments to follow.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+
+def generate_table(table_format: box.Box, headers: list[str], rows: Sequence[Union[tuple[Any, Any], list[Any]]],
+                   col_alignments: Optional[list[JustifyMethod]] = None, show_header: bool = True,
+                   row_styles: Optional[list[str]] = None) -> Table:
```
1. Type hints add some noise. When the definition becomes to long and needs to be wrapped, typically it becomes preferable to have one parameter per line. Especially so when some of the hints are long and complex.

    ```python
    def generate_table(
        table_format: box.Box,
        headers: list[str],
        rows: Sequence[Union[tuple[Any, Any], list[Any]]],
        col_alignments: Optional[list[JustifyMethod]] = None,
        show_header: bool = True,
        row_styles: Optional[list[str]] = None
    ) -> Table:
    ```

 2. I think the `tabulate.tabulate()` signature makes more sense.

    1. `table_format` should become a keyword parameter with a default value of `CLITheme.TABLE_FMT`.
    2. `headers` should also become a keyword parameter with a default value of `None`. It should be combined together with `show_header`: if set, then the arguments contains the header names; otherwise, don’t show the header at all.

 3. Maybe rows should be `Sequence[Sequence[Any]]`? `headers` should definitely be `Sequence[str]`. Probably others as well. Let’s discuss in person.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+                   col_alignments: Optional[list[JustifyMethod]] = None, show_header: bool = True,
+                   row_styles: Optional[list[str]] = None) -> Table:
+    """
+    Generates a rich Table object from given input rows.
```
From [PEP 257](https://peps.python.org/pep-0257/#multi-line-docstrings):

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+    table.row_styles = row_styles
+
+    if not col_alignments and len(rows) > 0:
+        # auto-detect on first row, numerical values on the right
```
From [PEP 8](https://peps.python.org/pep-0008/#comments):

> Comments should be complete sentences.

Please capitalise and end with a period.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+
+    if not col_alignments and len(rows) > 0:
+        # auto-detect on first row, numerical values on the right
+        col_alignments = ['right' if str(value).isnumeric() else 'left' for value in rows[0]]
```
You’re using `value` here and `elem` below. Probably better to stick to one, and likely `column` or `col` so that it ties well with the term ‘row’.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+        col_alignments = ['right' if str(value).isnumeric() else 'left' for value in rows[0]]
+    elif col_alignments and len(rows) > len(col_alignments):
+        while len(rows) > len(col_alignments):
+            col_alignments.append('left')
```
I don’t understand this. How is the size of `rows` related to the size of `col_alignments`?

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+            justify = col_alignments[i] if col_alignments[i] in ["left", "center", "right"] else "left"
+            table.add_column(header, overflow='fold', justify=justify)
+        else:
+            table.add_column(header, overflow='fold')
```
1. Both single and double quotes. Please pick one and apply it consistently.
 2. No need for the if-else structure. Calling `add_column()` has `justify='left'` as a default.

dchristidis replied:
I believe this still stands? `col_alignments` is, due to the previous block, filed with `['left', 'left', …]` unless set as an arguments. What prevents you from calling `table.add_column(header, overflow='fold', justify=justify)` unconditionally?

ericbanzuzi replied:
My initial thought process was that if the user accidentally provides a list `col_alignments` that is shorter than the elements in a row, the method could still fill the "missing" alignments with the default value `'left'` instead of breaking. E.g. you could have `row=['filled', 'test', 'row']` and `col_alignments=['left', 'right']` leading to `col_alignments=['left', 'right', 'left']`.

But yes I realised that you are right that if `col_alignments` will always be true since I always set it (if not provided) or fill the missing values. I will fix this.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,283 @@
+            table.add_column(header, overflow='fold')
+
+    for row in rows:
+        row = [elem if isinstance(elem, rich_widgets) and str(elem) != 'None' else str(elem) if elem is not None and str(elem) != 'None' else '' for elem in row]
```
This list comprehension with two conditional expressions appears complex and might require some simplification.

dchristidis reviewed (changes requested):
Some of my comments apply to your entire PR.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,285 @@
+        def time_formatter(timestamp: datetime) -> Text:
+            formatted_time = timestamp.strftime('%Y-%m-%d %H:%M:%S')
+            milliseconds = timestamp.microsecond // 1000
+            return Text(f'[{formatted_time},{milliseconds:03d}]')
```
What advantage does this have against `timestamp.isoformat(sep=' ', timespec='milliseconds')`?

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,285 @@
+
+    # Setting handler and formatter.
+    if not logger.handlers:
+        console = console or Console()
```
Aren’t ensuring that `console` is initialised inside `add_handler()`?

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,285 @@
+    :returns: CLI type
+    """
+    try:
+        cli_type = config_get('client', 'cli').lower()
```
Please use the `experimental` section.

dchristidis commented on bin/rucio:
```diff
@@ -293,8 +316,23 @@ def whoami_account(args):
+                console.print(table)
+        else:
+            console.width = sys.maxsize
+            console.print(table)
```
Isn’t this sequence of statements repeated multiple times? It looks to me like it should become a helper function.

dchristidis reviewed (changes requested):
There have been some changes merged affecting `rucio-admin` since you created your branch. Please rebase on top of the latest master.

ericbanzuzi commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+    *output: Any,
+    console: Console,
+    no_pager: bool = False
+) -> None:
```
The **Check Python Type Annotations** test currently fails because of the use of `Any` here. Not sure how to change it as `Console.print()` in `rich.console`  accepts `Any` as the type to the print statement too. 

Should I perhaps use `*output: object`  instead of `*output: Any`? That would lead to a passing test.

dchristidis replied:
No, I think you’re doing the right thing. The CI treats `Any` as missing a type, but there could be functions such as this where literally any type is permissible and handled properly.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+    col_alignments = col_alignments or ['right' if str(col).isnumeric() else 'left' for col in rows[0]]
+    headers = headers or [''] * len(rows[0])
+
+    if col_alignments and len(headers) > len(col_alignments):
```
The second statement is repeated below. The empty line should be removed. And doesn’t the first statement guarantee that `col_alignments` won’t be evaluating to false?

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+
+    headers = headers or [''] * len(rows[0])
+    for i, header in enumerate(headers):
+        table.add_column(header, overflow='fold', justify=col_alignments[i])
```
I wonder if `zip()` would look nicer here.

dchristidis commented on bin/rucio:
```diff
@@ -73,9 +81,14 @@ FAILURE = 1
 logger = logging.log
 gfal2_logger = logging.getLogger("gfal2")
 tablefmt = 'psql'
+cli_config = 'tabulate'
```
What is the purpose of this assignment?

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+) -> logging.Logger:
+    """
+    Factory method to set logger with handlers.
+
```
Please mention that it’s a copy of the `setup_logger()` with minor changes.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+        if cli_type not in ['rich', 'tabulate']:
+            cli_type = 'tabulate'
+    except (NoOptionError, NoSectionError, RucioException):
+        cli_type = 'tabulate'
```
I think it would be better to set the function’s `raise_exception=False`.

dchristidis commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,308 @@
+    try:
+        result = subprocess.run([default_pager], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+        less_available = result.returncode == 0
+    except (FileNotFoundError, Exception):
```
If you’re catching Exception, is there a benefit to mentioning `FileNotFoundError`?

dchristidis commented on bin/rucio:
```diff
@@ -1868,6 +2537,7 @@ def get_parser():
     oparser.add_argument('--robot', '-R', dest="human", default=True, action='store_false', help="All output in bytes and without the units. This output format is preferred by parsers and scripts.")
     oparser.add_argument('--user-agent', '-U', dest="user_agent", default='rucio-clients', action='store', help="Rucio User Agent")
     oparser.add_argument('--vo', dest="vo", metavar="VO", default=None, help="VO to authenticate at. Only used in multi-VO mode.")
+    oparser.add_argument("--no-pager", dest="no_pager", default=False, action='store_true', help="Do not pipe Rucio output into a pager.")
```
Use `help=argparse.SUPPRESS` for now.

dchristidis reviewed (dismissed):
This PR is too large to to check exhaustively, but I believe its approach in general is well-engineered. Any defects can more realistically be identified through actual use.

The failing CI test should be ignored (refer to [this](https://github.com/rucio/rucio/pull/7012#discussion_r1727686501) discussion for more information).

Given its magnitude, it would be ideal if the component leads also reviewed this.

voetberg commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,302 @@
+# Copyright European Organization for Nuclear Research (CERN) since 2012
```
This file should probably live in the same folders with the other client files

dchristidis replied:
@ericbanzuzi do you think you can find some time to address this one? I’m afraid that it is a blocker that I should have identified myself. The rest of @voetberg’s comments remain valid, but I feel like we can afford to work on them in the future.

ericbanzuzi replied:
@dchristidis yes, I think I can do it over the weekend or the weekend next week. I can also include some of the smallest and simplest fixes in these comments: [first](https://github.com/rucio/rucio/pull/7012#discussion_r1763294876), [second](https://github.com/rucio/rucio/pull/7012#discussion_r1763304438) and [third](https://github.com/rucio/rucio/pull/7012#discussion_r1763382976).

voetberg commented on bin/rucio:
```diff
@@ -1868,6 +2537,7 @@ def get_parser():
     oparser.add_argument('--robot', '-R', dest="human", default=True, action='store_false', help="All output in bytes and without the units. This output format is preferred by parsers and scripts.")
     oparser.add_argument('--user-agent', '-U', dest="user_agent", default='rucio-clients', action='store', help="Rucio User Agent")
     oparser.add_argument('--vo', dest="vo", metavar="VO", default=None, help="VO to authenticate at. Only used in multi-VO mode.")
+    oparser.add_argument("--no-pager", dest="no_pager", default=False, action='store_true', help=argparse.SUPPRESS)
```
Why does no_pager need to be included in the args? If it's being supressed, do we want people to be able to change this?

dchristidis replied:
The `--no-pager` is inspired by the respective command-line option in Git. The pager is used by default. This command-line option allows one to write to the terminal instead.

I explicitly asked for it to be suppressed from the help output because it’s ineffective when the output with Rich is not enabled (which is introduced as an experimental opt-in feature in this PR). It will be switched to visible with an appropriate help message when the Rich output becomes the default.

voetberg commented on bin/rucio:
```diff
@@ -1801,19 +2464,25 @@ def add_lifetime_exception(args):
-        logger.debug(traceback.format_exc())
+        if cli_config == 'rich':
+            if logger.level == logging.DEBUG:
+                logger.exception('Failure to submit exception. Please retry.')
```
Because this message is the same across all errors, it would make more sense to define the error message as a str var before the logic for picking the logger.

voetberg commented on bin/rucio:
```diff
@@ -1702,19 +2343,41 @@ def list_datasets_rse(args):
+            if args.long:
+                table_data = []
+                for dsn in client.list_datasets_per_rse(args.rse):
+                    table_data.append(['%s:%s' % (dsn['scope'], dsn['name']),
```
(This comment applies for all tables using %s formatting) Is there a reason for this not to be f-strings? Those are preferred for readability purposes.

ericbanzuzi replied:
No particular reason,  I reused the approach from the original code base.

voetberg commented on bin/rucio:
```diff
@@ -1670,26 +2282,55 @@ def list_account_usage(args):
-    usage = client.get_global_account_usage(account=args.usage_account)
-    for item in usage:
-        if (args.rse and args.rse in item['rse_expression']) or not args.rse:
+        if args.rse:
```
args.rse should default to None, so this logic is redundant, you can just pass args.rse whenever (ref: https://github.com/rucio/rucio/blob/e6097bd2ee4efb1ade8b4e262a28efdda6fb96a2/lib/rucio/client/accountclient.py#L330 )

voetberg commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,302 @@
+    rows: Sequence[Sequence[RenderableType]],
+    headers: Optional[Sequence[RenderableType]] = None,
+    row_styles: Optional[Sequence[StyleType]] = None,
+    col_alignments: Optional[list[JustifyMethod]] = None,
```
Is there a rhyme or reason to the left/right alignments? It may be cleaner to automatically assign that based on the first row's numeric/not numeric status for each column

ericbanzuzi replied:
I included a line, **231**,  that does automatic assignment for the simple cases 
```python
col_alignments = col_alignments or ['right' if str(col).isnumeric() else 'left' for col in rows[0]]
```

However, this approach does not currently work properly in all cases. For example:
- File sizes in human-readable formats (e.g., `100 MB`) are not handled correctly.
- If the first row in a column contains `None`, even though other rows are numerical, it will default to left alignment.
- For tables where columns have mixed value types (e.g., the two column key-value pair tables with different types), it's better to manually assign alignments.

Hence, the use of manual left/right alignments for now.

voetberg commented on lib/rucio/richclient.py:
```diff
@@ -0,0 +1,302 @@
+        console.print(*output)
+
+
+def get_cli_config() -> str:
```
Is this the only way to set if using rich or tabulate? Having an override in the args would be useful for people who want to try a command or two without changing their entire setup

dchristidis replied:
This was consciously done, as part of the release engineering. The first phase of testing will come from power users. We expect them to enable Rich in its entirety and, in a way, be forced to use it, not pick the output in each invocation.

voetberg replied:
Understandable, I just think being able to try it out without the commitment will increase adoption. That can come at a later date though if this is part of the release planning.

dchristidis replied:
Carrot vs stick. :laughing:

voetberg commented:
I largely agree with Dimintros' comments. Nothing looks objectively wrong, just a few style/nit-picks I'll work through. I'd love to see some examples and maybe a follow up PR in documentation explaining how to enable this and why you'd want to.

voetberg commented on bin/rucio:
```diff
@@ -73,9 +81,14 @@ FAILURE = 1
 DEFAULT_PORT = 80
 
+# Setting up rich console.
+console = Console(theme=Theme(CLITheme.LOG_THEMES), soft_wrap=True)
```
This setup, and the other setup in bin/rucio-admin, should probably be contained to a function in richclient and called within each function that uses the rich client to keep things more self-contained.

dchristidis commented:
> I'd love to see some examples and maybe a follow up PR in documentation explaining how to enable this and why you'd want to.

Please consult [Eric’s presentation](https://indico.cern.ch/event/1431526/#6-improving-the-rucio-cli-expe). It has many wonderful examples and the second-to-last slide describes how, if all goes well, this it to replace the existing implementation based on tabulate. Consequently, no additional documentation is necessary. If we time it with the new command-line structure that you’ve been working on, then we can update the existing documentation only once.

voetberg commented on bin/rucio:
```diff
@@ -327,36 +356,81 @@ def list_dataset_replicas(args):
-        dmeta = client.get_metadata(scope, name)
-        _fetch_datasets_for_meta(meta=dmeta)
+        if cli_config == 'rich':
+            result[dsn][replica['rse']] = [replica['rse'], str(replica['available_length']), str(replica['length'])]
```
Why can't this conversion be done when you're making the table? Seems like an unnecessary check

ericbanzuzi replied:
I suppose it could be done in the `generate_table` as well. I just wanted to limit the accepted elements in a row to `rich.console.RenderableType` (similarly to `rich.table.Table.add_rows()`), which does not include integer/float values.

voetberg commented on bin/rucio:
```diff
@@ -327,36 +356,81 @@ def list_dataset_replicas(args):
+            result[dsn][replica['rse']] = [replica['rse'], replica['available_length'], replica['length']]
+
+    if cli_config == 'rich':
+        status = Status('Fetching dataset replicas', spinner=CLITheme.SPINNER, spinner_style=CLITheme.SPINNER_STYLE, console=console)
```
Is there a way to define a placeholder for status when rich isn't enabled? There's a ton of repeated code between the if/else and that's rife for inconsistency problems.

ericbanzuzi replied:
Not sure about a placeholder but maybe switching to `status.start()` instead of the context manager might help with that. Then the if/else conditions would need to be in the core loops collecting the data since some of the logic between storing the rows for tabulate and rich are slightly different.

voetberg replied:
I think that's a better solution!

voetberg commented on bin/rucio:
```diff
@@ -384,75 +458,169 @@ def list_file_replicas(args):
-        return SUCCESS
+    if cli_config == 'rich':
+        status = Status('Fetching file replicas', spinner=CLITheme.SPINNER, spinner_style=CLITheme.SPINNER_STYLE, console=console)
+        with status:
```
[See this comment](https://github.com/rucio/rucio/pull/7012/files#r1763387604) - I won't replicate it for each part of the code where it applies but keep it in mind please.

voetberg commented:
> > I'd love to see some examples and maybe a follow up PR in documentation explaining how to enable this and why you'd want to.
> 
> Please consult [Eric’s presentation](https://indico.cern.ch/event/1431526/#6-improving-the-rucio-cli-expe). It has many wonderful examples and the second-to-last slide describes how, if all goes well, this it to replace the existing implementation based on tabulate. Consequently, no additional documentation is necessary. If we time it with the new command-line structure that you’ve been working on, then we can update the existing documentation only once.

Wow @ericbanzuzi This is a fantastic presentation!! I'm now thoroughly excited

Now that I see the release plan, I have no criticism. In theory I'd like if we had a better way to communicate experimental features like this to users outside these meetings but I think that's above my pay grade

ericbanzuzi commented:
I had some extra time, so on top what [I previously mentioned](https://github.com/rucio/rucio/pull/7012#discussion_r1809595953), I also tried to address [this](https://github.com/rucio/rucio/pull/7012#discussion_r1763387604) by reducing the amount of duplicate code using [this approach](https://github.com/rucio/rucio/pull/7012#discussion_r1763757708), and changed some of the %s formatting to f-strings ([see](https://github.com/rucio/rucio/pull/7012#discussion_r1763297884)).

ericbanzuzi committed:
Clients: Rich table output and logging with pager by default #6987

Key changes:
  * Introduce rich as a dependency
  * Introduce rich version of the cli that opted into use from the config file
  * Introduce rich tables
  * Introduce logging and exception handling with rich
  * Introduce highlighting of Rucio constants
  * Add hidden argument --no-pager to disable pager
  * Introduce spinners to most commands
  * Introduce lib/rucio/client/richclient.py with util functions for rich cli
  * Highlight list-file-replicas PFNs

Now only the simplest and fastest commands do not have a spinner to indicate it is running.
These commands are:
  rucio whoami / rucio-admin account info
  rucio list-rse-attributes / rucio-admin rse get-attribute
  rucio-admin account list-identities
  rucio rule-info
  rucio-admin rse get-distance

In addition, list-file-replicas PFNs in --no-pager are hyperlinks if terminal supports OSC 8.

voetberg reviewed (approved):
All the changes look good to me! Thank you for addressing my comments. 

After convo with @bari12 and @dchristidis, we decided this will probably have to be merged after #6923, but I can do that rebase myself, so as it exists now, this seems good!

bari12 commented:
@voetberg I think we should do the other way around (Maybe this got lost in our chat). Merge Erics first and rebase your PR to it.
